lib/CL/devices/basic/basic.c

```diff
@@ -855,20 +889,130 @@ pocl_basic_svm_alloc (cl_device_id dev, cl_svm_mem_flags flags, size_t size)
+                          CL_INVALID_OPERATION,
+                          "Device does not support Shared USM allocations\n");
+      break;
+    default:
+      POCL_GOTO_ERROR_ON (1, CL_INVALID_PROPERTY,
+                          "Unknown USM AllocType requested\n");
+    }
+
+  ptr = pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, size);
+
+  if (ptr != NULL)
+    {
+      pocl_basic_usm_allocation_t *all
+          = malloc (sizeof (pocl_basic_usm_allocation_t));
+      POCL_GOTO_ERROR_COND ((all == NULL), CL_OUT_OF_HOST_MEMORY);
+      all->ptr = ptr;
+      all->size = size;
+      all->flags = flags;
+      all->alloc_type = alloc_type;
+      POCL_LOCK (usm_lock);
+      DL_APPEND (usm_allocations, all);
+      POCL_UNLOCK (usm_lock);
+
+      POCL_MSG_WARN ("ALLOCATED: ALL_PTR: %p ALL_SIZE: %zu\n", all->ptr,
+                     all->size);
+    }
+
+ERROR:
+  if (err_code)
+    *err_code = errcode;
+
+  return ptr;
+}
+
+void
+pocl_basic_usm_free (cl_device_id dev, void *svm_ptr)
+{
+  POCL_LOCK (usm_lock);
+  pocl_basic_usm_allocation_t *item = NULL, *tmp = NULL;
+  DL_FOREACH_SAFE (usm_allocations, item, tmp)
+  {
+    if (item->ptr == svm_ptr)
+      {
+        DL_DELETE (usm_allocations, item);
+        free (item);
+        break;
+      }
+  }
+  POCL_UNLOCK (usm_lock);
+
+  /* we should always find it, b/c there are checks
+   * in the runtime API (clMemFreeINTEL) */
+  assert (item != NULL);
+  pocl_aligned_free (svm_ptr);
+}
+
 void
 pocl_basic_svm_copy (cl_device_id dev, void *__restrict__ dst,
                      const void *__restrict__ src, size_t size)
 {
   memcpy (dst, src, size);
 }
 
+static cl_bitfield
+pocl_basic_get_mem_caps (cl_device_id device, cl_device_info Type)
+{
+  switch (Type)
+    {
+    case CL_DEVICE_HOST_MEM_CAPABILITIES_INTEL:
+      return device->host_usm_capabs;
+    case CL_DEVICE_DEVICE_MEM_CAPABILITIES_INTEL:
+      return device->device_usm_capabs;
+    case CL_DEVICE_SINGLE_DEVICE_SHARED_MEM_CAPABILITIES_INTEL:
+      return device->single_shared_usm_capabs;
+    case CL_DEVICE_CROSS_DEVICE_SHARED_MEM_CAPABILITIES_INTEL:
+      return device->cross_shared_usm_capabs;
+    case CL_DEVICE_SHARED_SYSTEM_MEM_CAPABILITIES_INTEL:
+      return device->system_shared_usm_capabs;
+    default:
+      assert (0 && "unhandled switch value");
+    }
+  return 0;
+}
+
 cl_int
 pocl_basic_get_device_info_ext (cl_device_id device, cl_device_info param_name,
                             size_t param_value_size, void *param_value,
                             size_t *param_value_size_ret)
 {
   switch (param_name)
     {
+    case CL_DEVICE_HOST_MEM_CAPABILITIES_INTEL:
+    case CL_DEVICE_DEVICE_MEM_CAPABILITIES_INTEL:
+    case CL_DEVICE_SINGLE_DEVICE_SHARED_MEM_CAPABILITIES_INTEL:
+    case CL_DEVICE_CROSS_DEVICE_SHARED_MEM_CAPABILITIES_INTEL:
+    case CL_DEVICE_SHARED_SYSTEM_MEM_CAPABILITIES_INTEL:
+      {
+        cl_bitfield Caps = pocl_basic_get_mem_caps (device, param_name);
```

Comment:
Lower case "caps" is the correct C style?

Reply:
fixed